REGRESSION(287399@main): [Skia] Broke rendering of various PDFs

nikolaszimmermann opened the pull request:
#### 48db7480703cf267922ad0b91f1ef965d8fa64a4
<pre>
REGRESSION(287399@main): [Skia] Broke rendering of various PDFs
<a href="https://bugs.webkit.org/show_bug.cgi?id=295538">https://bugs.webkit.org/show_bug.cgi?id=295538</a>

Reviewed by Carlos Garcia Campos.

Fix PDF rendering regression caused by cross-thread transfer of accelerated
ImageBitmap. When a texture-backed image is transferred across threads, the
GPU texture may not be valid for the current Skia context.

The fix has two parts:

1. Flush GPU commands before cross-thread transfer by calling
   flushDrawingContext() in OffscreenCanvasRenderingContext2D::transferToImageBuffer().
   This is implemented via ImageBufferSkiaAcceleratedBackend::flushContext()
   which creates a GL fence and performs a server wait.

2. Rewrap textures at draw time in GraphicsContextSkia::drawNativeImage().
   When drawing a texture-backed image, check if it&apos;s valid for the current
   GrContext using isValid(). If not (because it was created in a different
   thread/context), use SkImages::BorrowTextureFrom() to create a new SkImage
   wrapper around the existing backend texture. Before rewrapping, create
   a GL fence and server wait to ensure pending GPU operations on the source
   image are complete.

This approach simplifies cross-thread transfer by removing the explicit
prepare/finalize pattern from ImageBitmap (prepareForCrossThreadTransfer,
finalizeCrossThreadTransfer, m_fence member), ImageBuffer
(sinkIntoImageBufferForCrossThreadTransfer,
sinkIntoImageBufferAfterCrossThreadTransfer), and SerializedScriptValue.

Also refactor the fence creation API: rename the implementation to
createAcceleratedRenderingFenceInternal, add public static
createAcceleratedRenderingFence() methods for both SkSurface* and
sk_sp&lt;SkImage&gt;, and remove the private createAcceleratedRenderingFenceIfNeeded
methods by inlining the null/texture-backed checks into callers.

Covered by existing tests.

* Source/WebCore/bindings/js/SerializedScriptValue.cpp:
(WebCore::CloneSerializer::dumpImageBitmap):
(WebCore::CloneDeserializer::readTransferredImageBitmap):
* Source/WebCore/html/ImageBitmap.cpp:
(WebCore::ImageBitmap::prepareForCrossThreadTransfer): Deleted.
(WebCore::ImageBitmap::finalizeCrossThreadTransfer): Deleted.
* Source/WebCore/html/ImageBitmap.h:
* Source/WebCore/html/canvas/OffscreenCanvasRenderingContext2D.cpp:
(WebCore::OffscreenCanvasRenderingContext2D::transferToImageBuffer):
* Source/WebCore/platform/graphics/ImageBuffer.cpp:
(WebCore::DefaultSerializedImageBuffer::DefaultSerializedImageBuffer):
(WebCore::ImageBuffer::sinkIntoImageBufferForCrossThreadTransfer): Deleted.
(WebCore::ImageBuffer::sinkIntoImageBufferAfterCrossThreadTransfer): Deleted.
* Source/WebCore/platform/graphics/ImageBuffer.h:
* Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
(WebCore::GraphicsContextSkia::drawNativeImage):
(WebCore::createAcceleratedRenderingFenceInternal):
(WebCore::GraphicsContextSkia::createAcceleratedRenderingFence):
(WebCore::GraphicsContextSkia::trackAcceleratedRenderingFenceIfNeeded):
(WebCore::createAcceleratedRenderingFence): Deleted.
(WebCore::GraphicsContextSkia::createAcceleratedRenderingFenceIfNeeded): Deleted.
* Source/WebCore/platform/graphics/skia/GraphicsContextSkia.h:
* Source/WebCore/platform/graphics/skia/ImageBufferSkiaAcceleratedBackend.cpp:
(WebCore::ImageBufferSkiaAcceleratedBackend::flushContext):
* Source/WebCore/platform/graphics/skia/ImageBufferSkiaAcceleratedBackend.h:

Canonical link: <a href="https://commits.webkit.org/304687@main">https://commits.webkit.org/304687@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/887da01fb8ac1ff52c7a54a399aecfc27d140f6a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/136280 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/8637 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/47560 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/143991 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/89251 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/138151 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/9316 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/8481 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/104205 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/89251 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/139225 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/6778 "Passed tests") | [❌ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/122119 "Found 1 new API test failure: TestWebKitAPI.WKUserContentController.DidAssociateFormControlsFromShadowTree (failure)") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/85038 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/6431 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/4095 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/4583 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/115722 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/40320 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/146735 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/8319 "Built successfully") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/40888 "Found 1 new test failure: fast/images/mac/play-all-pause-all-animations-context-menu-items.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/112544 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/8336 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/6993 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/112888 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28652 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/6364 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/118424 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/62306 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/8367 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/36481 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/8085 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/71926 "Built successfully") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/8307 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/8159 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

carlosgcampos commented on Source/WebCore/bindings/js/SerializedScriptValue.cpp:
```diff
@@ -1668,9 +1668,6 @@ class CloneSerializer : public CloneBase {
         Ref imageBitmap = jsCast<JSImageBitmap*>(obj)->wrapped();
         auto index = m_transferredImageBitmaps.find(obj);
         if (index != m_transferredImageBitmaps.end()) {
-#if USE(SKIA)
-            imageBitmap->prepareForCrossThreadTransfer();
-#endif
```
This is great, I never liked this prepare/finalize.

nikolaszimmermann committed:
REGRESSION(287399@main): [Skia] Broke rendering of various PDFs
https://bugs.webkit.org/show_bug.cgi?id=295538

Reviewed by Carlos Garcia Campos.

Fix PDF rendering regression caused by cross-thread transfer of accelerated
ImageBitmap. When a texture-backed image is transferred across threads, the
GPU texture may not be valid for the current Skia context.

The fix has two parts:

1. Flush GPU commands before cross-thread transfer by calling
   flushDrawingContext() in OffscreenCanvasRenderingContext2D::transferToImageBuffer().
   This is implemented via ImageBufferSkiaAcceleratedBackend::flushContext()
   which creates a GL fence and performs a server wait.

2. Rewrap textures at draw time in GraphicsContextSkia::drawNativeImage().
   When drawing a texture-backed image, check if it's valid for the current
   GrContext using isValid(). If not (because it was created in a different
   thread/context), use SkImages::BorrowTextureFrom() to create a new SkImage
   wrapper around the existing backend texture. Before rewrapping, create
   a GL fence and server wait to ensure pending GPU operations on the source
   image are complete.

This approach simplifies cross-thread transfer by removing the explicit
prepare/finalize pattern from ImageBitmap (prepareForCrossThreadTransfer,
finalizeCrossThreadTransfer, m_fence member), ImageBuffer
(sinkIntoImageBufferForCrossThreadTransfer,
sinkIntoImageBufferAfterCrossThreadTransfer), and SerializedScriptValue.

Also refactor the fence creation API: rename the implementation to
createAcceleratedRenderingFenceInternal, add public static
createAcceleratedRenderingFence() methods for both SkSurface* and
sk_sp<SkImage>, and remove the private createAcceleratedRenderingFenceIfNeeded
methods by inlining the null/texture-backed checks into callers.

Covered by existing tests.

* Source/WebCore/bindings/js/SerializedScriptValue.cpp:
(WebCore::CloneSerializer::dumpImageBitmap):
(WebCore::CloneDeserializer::readTransferredImageBitmap):
* Source/WebCore/html/ImageBitmap.cpp:
(WebCore::ImageBitmap::prepareForCrossThreadTransfer): Deleted.
(WebCore::ImageBitmap::finalizeCrossThreadTransfer): Deleted.
* Source/WebCore/html/ImageBitmap.h:
* Source/WebCore/html/canvas/OffscreenCanvasRenderingContext2D.cpp:
(WebCore::OffscreenCanvasRenderingContext2D::transferToImageBuffer):
* Source/WebCore/platform/graphics/ImageBuffer.cpp:
(WebCore::DefaultSerializedImageBuffer::DefaultSerializedImageBuffer):
(WebCore::ImageBuffer::sinkIntoImageBufferForCrossThreadTransfer): Deleted.
(WebCore::ImageBuffer::sinkIntoImageBufferAfterCrossThreadTransfer): Deleted.
* Source/WebCore/platform/graphics/ImageBuffer.h:
* Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
(WebCore::GraphicsContextSkia::drawNativeImage):
(WebCore::createAcceleratedRenderingFenceInternal):
(WebCore::GraphicsContextSkia::createAcceleratedRenderingFence):
(WebCore::GraphicsContextSkia::trackAcceleratedRenderingFenceIfNeeded):
(WebCore::createAcceleratedRenderingFence): Deleted.
(WebCore::GraphicsContextSkia::createAcceleratedRenderingFenceIfNeeded): Deleted.
* Source/WebCore/platform/graphics/skia/GraphicsContextSkia.h:
* Source/WebCore/platform/graphics/skia/ImageBufferSkiaAcceleratedBackend.cpp:
(WebCore::ImageBufferSkiaAcceleratedBackend::flushContext):
* Source/WebCore/platform/graphics/skia/ImageBufferSkiaAcceleratedBackend.h:

Canonical link: https://commits.webkit.org/304687@main
